rust/kernel/user_ptr.rs

```diff
@@ -111,6 +110,12 @@ impl UserSlicePtr {
     }
 }
 
+impl Clone for UserSlicePtr {
```

Comment:
I don't think this is safe -- it lets you read the same address 2x which ends badly (e.g. https://www.usenix.org/conference/usenixsecurity17/technical-sessions/presentation/wang-pengfei)

Reply:
While this is something we ideally would like Rust to help address, it isn't really a memory safety issue: it *is* safe to read multiple times from the same location, it is desirable to avoid it because it may lead to correctness issues.

It turns out that I use this in 3 places in binder: in all 3 I first read some data, do some work, then write it back. So I can ditch this and replace it with a `reader_writer` function.

However, I still need a public constructor for `UserSlicePtr` because in one of the ioctls I get pointers from a struct I read from userspace. And having a public constructor still allows driver writers to 'double-fetch'. I added a paragraph about TOCTOU bugs in the constructor.

Let me know if you have other ideas on how we can approach this.

Reply:
The `reader_writer` seems much better.

In terms of how to do `ioctl`, IIRC the kernel API is an `int` opcode + a `long` value. How are you structuring that in your `ioctl` API, just passing a `usize` to the callback and letting it cast it? I might suggest passing a newtype, instead of a raw `usize` and giving it an `as_ptr()` method or something like that.

Reply:
I'm passing just a `usize` because they may just be a `usize`. However, there was an attempt at standardising `ioctl` opcodes, so they carry information about direction (none/read/write/read-write) and size of the data being transferred. In these cases, I actually provide a `dispatch_ioctl` that interprets these bits and calls the appropriate function. Here's what it looks like (I'm in the process of cleaning it up and will soon send a PR for it):

```
pub fn dispatch_ioctl<T: IoctlHandler>(
    handler: &T::SelfType,
    file: &File,
    cmd: u32,
    arg: usize,
) -> KernelResult<i64> {
    let dir = (cmd >> bindings::_IOC_DIRSHIFT) & bindings::_IOC_DIRMASK;
    if dir == bindings::_IOC_NONE {
        return T::pure(handler, cmd, arg);
    }

    let size = (cmd >> bindings::_IOC_SIZESHIFT) & bindings::_IOC_SIZEMASK;
    let data = unsafe { UserSlicePtr::new(arg as _, size as _) }?;
    const READ_WRITE: u32 = bindings::_IOC_READ | bindings::_IOC_WRITE;
    match dir {
        bindings::_IOC_WRITE => T::write(handler, cmd, &mut data.reader()),
        bindings::_IOC_READ => T::read(handler, cmd, &mut data.writer()),
        READ_WRITE => T::read_write(handler, file, cmd, data), 
        _ => Err(Error::EINVAL),
    }
}
```

So drivers that conform to this (binder mostly does, there is one exception), can just deal with UserSlicePtr[Reader|Writer], which is great.

Binder has a conforming case that still requires a constructor though, it's [`BINDER_WRITE_READ`](https://elixir.bootlin.com/linux/latest/source/include/uapi/linux/android/binder.h#L220) -- the `ioctl` carries a pointer to [`binder_write_read`](https://elixir.bootlin.com/linux/latest/source/include/uapi/linux/android/binder.h#L176), which itself has pointers to additional data that the driver must read and write.